timApp/modules/cs/js/editor/math-editor/formula-editor.component.ts

```diff
@@ -0,0 +1,1072 @@
+     */
+    async handleFormulaCancel() {
+        const oldContent =
+            this.oldContent.before +
+            this.oldContent.editing +
+            this.oldContent.after;
+        // content hasn't changed from what it was before opening formula editor
+        // so cancel
+        if (
+            oldContent === this.editor.content ||
+            (await showConfirm(
+                $localize`Are you sure?`,
+                $localize`This will clear the editor.`
+            ))
+        ) {
+            // cancelling creation of a new formula
+            if (this.oldContent.editing === "") {
+                const endPos = this.oldContent.before.length;
+                this.cancelClose.emit(endPos);
+            } else {
+                // cancelling editing formula
+                const endPos =
+                    this.oldContent.before.length +
+                    this.oldContent.editing.length;
+                this.cancelClose.emit(endPos);
+            }
+
+            this.clearFields();
+
+            this.editor.content = oldContent;
+        }
+    }
+
+    /**
+     * Moves cursor to where cursor symbol is
+     * and deletes the cursor symbol.
+     * @param activeField field being edited
+     */
+    setMathQuillCursor(activeField: FormulaFieldComponent) {
+        const span = activeField.visualInput.nativeElement;
+        const cursor = CURSOR;
+        const children = span.getElementsByTagName("span");
+        for (const child of children) {
+            // try to pick the correct element to click
+            // textContent comparison is not enough to find
+            // the unique, correct element to click
+            if (
+                child.textContent === cursor &&
+                child.hasAttribute("mathquill-command-id") &&
+                !child.classList.contains("mq-non-leaf")
+            ) {
+                // clicks at position where cursor is
+                child.dispatchEvent(
+                    new MouseEvent("mousedown", {
+                        bubbles: true,
+                    })
+                );
+                child.dispatchEvent(
+                    new MouseEvent("mouseup", {
+                        bubbles: true,
+                    })
+                );
+                // removes cursor symbol
+                activeField.mathField.keystroke("Right Backspace");
+                activeField.mathField.focus();
+                return;
+            }
+        }
+
+        // put focus to field even if it doesn't have cursor symbol
+        activeField.mathField.focus();
+    }
+
+    /**
+     * Adds formula to both fields in last known cursor position.
+     * @param formulaInput LaTeX-formula to be added to fields.
+     */
+    addFormula(formulaInput: FormulaEvent) {
+        const cursorPosition = formulaInput.text.indexOf(CURSOR);
+        const formulaWithoutCursor = formulaInput.text.replace(CURSOR, "");
```

Comment:
Oliko jokin erityinen syy käsitellä liittämistä MathQuilliin näin? MathQuill osaa automaattisesti käsitellä tuon kursorin siirtämistä (katso esim. Abitti) ja lisäksi osaa jopa ympäröidä kaavat sopivasti. Yksi tapa voisi siis olla vaan, että MathQuillin fieldiin kirjoitetaan tuo `formulaWithoutCursor`. Kursori oli alun perin tarkoitettu vain ACE- ja tekstieditorille, jossa MathQuillin kaltaista fiksua ympäröintilogiikkaa ei ole.

Toinen vaihtoehto olisi toki toteuttaa tuo MathQuillin ympäräintilogiikkaa itse, mutta taitaa mennä vähän liian vaikeaksi :)

Reply:
Haluttiin hyödyntää kursoria, että voi määrittää omia painikkeita ja saada kursorin siihen kohtaan, mihin haluaa. Muikussa ja digabissa painikkeet on toteutettu nin että joutuu määrittämään jokaista painiketta kohden mathquillin sisäisiä asetuksia, joka ei ole kovin kätevää käyttäjän näkökulmasta.